cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/inference/DFGFunctionSummaries.kt

```diff
@@ -0,0 +1,197 @@
+            for (typeEntry in typeEntryList) {
+                if (typeEntry.first == mostPreciseType) {
+                    mostPreciseClassEntries.add(typeEntry.second)
+                } else if (typeEntry.first in superTypes) {
+                    mostPreciseClassEntries.clear()
+                    mostPreciseClassEntries.add(typeEntry.second)
+                    mostPreciseType = typeEntry.first
+                    superTypes = getAllSupertypes(typeEntry.first)
+                }
+            }
+            if (mostPreciseClassEntries.size > 1) {
+                mostPreciseClassEntries.sortByDescending { it.signature?.size ?: 0 }
+            }
+            functionToDFGEntryMap[matchingEntries.first()]
+        } else {
+            null
+        }
+    }
+
+    private fun getAllSupertypes(type: Type?): Set<Type> {
+        if (type == null) return setOf()
+        val superTypes = type.superTypes
+        superTypes.addAll(type.superTypes.flatMap { getAllSupertypes(it) })
+        return superTypes
+    }
+
+    private fun applyDfgEntryToFunctionDeclaration(
+        functionDeclaration: FunctionDeclaration,
+        dfgEntries: List<DFGEntry>
+    ) {
+        for (entry in dfgEntries) {
+            val from =
+                if (entry.from.startsWith("param")) {
+                    try {
+                        val paramIndex = entry.from.removePrefix("param").toInt()
+                        functionDeclaration.parameters[paramIndex]
+                    } catch (e: NumberFormatException) {
+                        null
+                    }
+                } else if (entry.from == "base") {
+                    (functionDeclaration as? MethodDeclaration)?.receiver
+                } else {
+                    null
+                }
+            val to =
+                if (entry.to.startsWith("param")) {
+                    try {
+                        val paramIndex = entry.to.removePrefix("param").toInt()
+                        functionDeclaration.parameters[paramIndex]
+                    } catch (e: NumberFormatException) {
+                        null
+                    }
+                } else if (entry.to == "base") {
+                    (functionDeclaration as? MethodDeclaration)?.receiver
+                } else if (entry.to == "return") {
+                    functionDeclaration
+                } else if (entry.to.startsWith("return")) {
+                    val returnIndex = entry.to.removePrefix("param").toInt()
+                    // TODO: It would be nice if we could model the index. Not sure how this is done
+                    functionDeclaration
+                } else {
+                    null
+                }
+            // TODO: It would make sense to model properties here. Could be the index of a return
+            // value, full vs. partial flow or whatever comes to our minds in the future
+            to?.let { from?.addNextDFG(it) }
```

Comment:
So this only connects the flow *inside* the function declaration, but not back to the call expression, correct?

So if I have a call like `foo(a, b)` and model a flow from param0 to param1, I will see a flow from one param declaration to the other, but I won't see a flow at the call site from `a`, to `b`?

Reply:
This is correct and consistent with handling any other function declaration. I'll think about a solution for this issue

Reply:
I proposed a solution for this issue. The idea is that we save "writes" to a parameter (i.e., when it receives an incoming DFG edge) in a map in the `DFGSummaries` object. We also collect all function calls to inferred functions in the standard `DFGPass`. After having iterated the whole graph, we check the calls to the collected functions again, look up the parameters which flow out of the function declaration and the parameters to the arguments and draw these edges. The edges between the arguments and the parameters are labeled with the `CALLING_CONTEXT_IN` and `CALLING_CONTEXT_OUT` to enable context sensitivity in this case.

The only downside of this approach is that neo4j-ogm requires a custom serialization. I used `toString()` for now but we could also rely on the `id` instead. For some reason, the neo4j test still fails and I'm not sure why :/